ext/ffi/lib.rs

```diff
@@ -247,6 +247,32 @@ enum NativeType {
   F64,
   Pointer,
   Function,
+  Struct(Vec<NativeType>),
+}
+
+impl NativeType {
+  fn get_size(&self) -> usize {
+    match self {
+      NativeType::Void => 0,
+      NativeType::U8 | NativeType::I8 => 1,
+      NativeType::U16 | NativeType::I16 => 2,
+      NativeType::U32 | NativeType::I32 | NativeType::F32 => 4,
+      NativeType::U64 | NativeType::I64 | NativeType::F64 => 8,
+      NativeType::USize
+      | NativeType::ISize
+      | NativeType::Pointer
+      | NativeType::Function => std::mem::size_of::<usize>(),
+      NativeType::Struct(fields) => NativeType::compute_struct_size(fields),
+    }
+  }
+
+  fn compute_struct_size(fields: &[NativeType]) -> usize {
+    let mut size = 0;
+    for field in fields {
+      size += field.get_size();
+    }
+    size
+  }
```

Comment:
Doesn't seem to account for padding, i.e. its assuming a packed representation, can you confirm?

Reply:
If it's intended, a note or renaming might be preferred, but I'm fairly doubtful that this is correct, and would prefer test cases covering objects with padding.

Reply:
Very much correct, this expects packed representation. Essentially, it's up to the user to define any packing bytes if needed.

Reply:
Many users may not know about padding, or know how to compute it; if possible, it may be preferable to handle padding for consumers...

Reply:
Users kind of will need to know about it to use structs in Deno FFI.

eg. Imagine a `["u8", "u32"]` struct. If I'm not badly mistaken, this will essentially become `["u8", "u8" x 3 padding, "u32"]` if not packed, right? Now, if we compute the padding for the user we'll then expect an 8 byte buffer. (Side note: Users would need some way to opt out of padding for packed structs as well.)

Now, a user who does not understand padding will read `"u8"´ + `"u32"`, create a 5 byte array and pass it along. ERROR! Horrible things happen! Either we check and throw an error about the buffer being too small (good handling of this), or the receiving native library writes beyond the buffer's limits and memory gets corrupted.

Since we do not offer a way to create custom buffers from structs definition, nor do we offer any help in accessing the proper parts of it (eg. in the above case the padding bytes at index 1, 2 and 3 shouldn't really even be offered for reading), it doesn't really make sense to me to try and be automatic about padding handling either.

Reply:
No real opinion from myself :shrug:.
> (Side note: Users would need some way to opt out of padding for packed structs as well.)

I believe that this would make the most sense. Nonetheless, I believe documenting the implemented behaviour is all that would suffice.

Reply:
I think this topic would be more easily discussed if framed as a 2-step decision:
1. should the struct definition be understood as the declaration of the struct fields, or the definition of its memory layout?
2. if the answer is the former (struct declaration), should this PR be actually adding support for packed structs, or padded structs?

If I understand both positions correctly, @aapoalas is arguing that the answer to 1. should be the memory layout, while @Phosra understands it as the struct declaration. Am I right?

Reply:
Turns out the discussion is even simpler: How does libffi actually work with structs? The answer is that it takes in the struct fields declaration, and not a memory layout definition. So my opinion is just plain unworkable from the libffi usage viewpoint.

libffi doesn't seem to offer any way to input packed structs, so I think that gets left as an exercise to the users to figure out... Maybe it would just be equal to a bunch of usizes in a struct, or such.

Reply:
padded struct field declaration it is, then.

Reply:
> libffi doesn't seem to offer any way to input packed structs, so I think that gets left as an exercise to the users to figure out... Maybe it would just be equal to a bunch of usizes in a struct, or such.

I expect that it would just be giving the byte count as the amount of bytes in the struct (what the current code in question computes), treating it as a blob, and manual extraction/compression into the packed representation. (Could be risky and might break against some C compilers' layouts?)